File: client/src/lib/api/checkout.ts
/**
 * Checkout API client
 */

import { api } from './index';

export interface PassengerInfo {
    nid: string;
    name: string;
    seat_id: string;
    seat_number: string;
    date_of_birth: string;
    gender: string;
    phone: string;
    email: string;
    price?: number;
}

export interface CreateCheckoutRequest {
    trip_id: string;
    hold_id: string;
    from_station_id: string;
    to_station_id: string;
    passengers: PassengerInfo[];
    coupon_code?: string;
}

export interface CheckoutSession {
    id: string;
    trip_id: string;
    hold_id: string;
    passengers: PassengerInfo[];
    base_price_paisa: number;
    tax_paisa: number;
    booking_fee_paisa: number;
    discount_paisa: number;
    total_paisa: number;
    currency: string;
    coupon_code?: string;
    coupon_discount: number;
    coupon_validated: boolean;
    payment_method: string;
    status: 'pending' | 'in_progress' | 'confirmed' | 'failed' | 'expired';
    order_id?: string;
    created_at: string;
    expires_at: string;
}

export interface HoldStatus {
    hold_id: string;
    trip_id: string;
    from_station_id: string;
    to_station_id: string;
    held_seat_ids: string[];
    expires_at: string;
    total_paisa: number;
    status: string;
}

export const checkoutApi = {
    createCheckout: async (payload: CreateCheckoutRequest): Promise<CheckoutSession> => {
        const response = await api.post<CheckoutSession>('/v1/checkout', payload);
        return response;
    },

    getCheckout: async (checkoutId: string): Promise<CheckoutSession> => {
        const response = await api.get<CheckoutSession>(`/v1/checkout/${checkoutId}`);
        return response;
    },

    getCheckoutByHoldId: async (holdId: string): Promise<CheckoutSession> => {
        const response = await api.get<CheckoutSession>(`/v1/checkout/hold/${holdId}`);
        return response;
    },

    updateCheckout: async (checkoutId: string, payload: Partial<CreateCheckoutRequest>): Promise<CheckoutSession> => {
        const response = await api.patch<CheckoutSession>(`/v1/checkout/${checkoutId}`, payload);
        return response;
    },

    confirmCheckout: async (checkoutId: string): Promise<CheckoutSession> => {
        const response = await api.post<CheckoutSession>(`/v1/checkout/${checkoutId}/confirm`, {});
        return response;
    },

    listCheckouts: async (params?: { status?: string; limit?: number; offset?: number }): Promise<{ sessions: CheckoutSession[]; total: number }> => {
        const query = new URLSearchParams();
        if (params?.status) query.append('status', params.status);
        if (params?.limit) query.append('limit', params.limit.toString());
        if (params?.offset) query.append('offset', params.offset.toString());
        
        const response = await api.get<{ sessions: CheckoutSession[]; total: number }>(`/v1/checkout?${query.toString()}`);
        return response;
    },
};

File: server/services/order/internal/domain/checkout.go
package domain

import (
	"time"
)

// CheckoutSession represents a checkout session created from a seat hold
type CheckoutSession struct {
	ID             string             `json:"id" gorm:"primaryKey"`
	OrganizationID string             `json:"organization_id" gorm:"index"`
	UserID         string             `json:"user_id" gorm:"index"`
	TripID         string             `json:"trip_id"`
	HoldID         string             `json:"hold_id" gorm:"uniqueIndex"`
	FromStationID  string             `json:"from_station_id"`
	ToStationID    string             `json:"to_station_id"`

	// Passenger details (filled during checkout)
	Passengers     []CheckoutPassenger `json:"passengers" gorm:"foreignKey:CheckoutSessionID"`

	// Pricing
	BasePricePaisa    int64            `json:"base_price_paisa"`
	TaxPaisa          int64            `json:"tax_paisa"`
	BookingFeePaisa   int64            `json:"booking_fee_paisa"`
	DiscountPaisa     int64            `json:"discount_paisa"`
	TotalPaisa        int64            `json:"total_paisa"`
	Currency          string           `json:"currency"`

	// Coupon
	CouponCode        string           `json:"coupon_code"`
	CouponDiscount    int64            `json:"coupon_discount"`
	CouponValidated   bool             `json:"coupon_validated"`

	// Payment
	PaymentMethod     string           `json:"payment_method"`
	PaymentGateway    string           `json:"payment_gateway"`

	// Status
	Status            CheckoutStatus   `json:"status" gorm:"index"`

	// Order reference (set after checkout confirmation)
	OrderID           string           `json:"order_id" gorm:"index"`

	// Timestamps
	CreatedAt         time.Time        `json:"created_at"`
	UpdatedAt         time.Time        `json:"updated_at"`
	ExpiresAt         time.Time        `json:"expires_at"`
	CompletedAt       *time.Time       `json:"completed_at,omitempty"`
}

// CheckoutStatus represents the lifecycle of a checkout session
type CheckoutStatus string

const (
	CheckoutStatusPending    CheckoutStatus = "pending"
	CheckoutStatusInProgress CheckoutStatus = "in_progress"
	CheckoutStatusConfirmed  CheckoutStatus = "confirmed"
	CheckoutStatusFailed     CheckoutStatus = "failed"
	CheckoutStatusExpired    CheckoutStatus = "expired"
)

// CheckoutPassenger represents a passenger in a checkout session
type CheckoutPassenger struct {
	ID                    uint   `json:"id" gorm:"primaryKey"`
	CheckoutSessionID     string `json:"-" gorm:"index"`
	NID                   string `json:"nid"`
	Name                  string `json:"name"`
	SeatID                string `json:"seat_id"`
	SeatNumber            string `json:"seat_number"`
	SeatClass             string `json:"seat_class"`
	SeatType              string `json:"seat_type"`
	DateOfBirth           string `json:"date_of_birth"`
	Gender                string `json:"gender"`
	Age                   int    `json:"age"`
	Phone                 string `json:"phone"`
	Email                 string `json:"email"`
	EmergencyContactName  string `json:"emergency_contact_name"`
	EmergencyContactPhone string `json:"emergency_contact_phone"`
}

// PricingBreakdown represents the pricing calculation result
type PricingBreakdown struct {
	BasePricePaisa    int64            `json:"base_price_paisa"`
	TaxPaisa          int64            `json:"tax_paisa"`
	BookingFeePaisa   int64            `json:"booking_fee_paisa"`
	DiscountPaisa     int64            `json:"discount_paisa"`
	TotalPaisa        int64            `json:"total_paisa"`
	Currency          string           `json:"currency"`
	AppliedRules      []PricingRule    `json:"applied_rules,omitempty"`
	CouponApplied     bool             `json:"coupon_applied"`
	CouponDiscount    int64            `json:"coupon_discount"`
	CouponCode        string           `json:"coupon_code,omitempty"`
}

// PricingRule represents an applied pricing rule
type PricingRule struct {
	RuleID     string  `json:"rule_id"`
	RuleName   string  `json:"rule_name"`
	Multiplier float64 `json:"multiplier"`
}

// CouponValidation represents the result of coupon validation
type CouponValidation struct {
	Valid          bool   `json:"valid"`
	DiscountPaisa  int64  `json:"discount_paisa"`
	Message        string `json:"message"`
	CouponCode     string `json:"coupon_code"`
	DiscountType   string `json:"discount_type"`
	DiscountValue  float64 `json:"discount_value"`
}
